arduino/waveform/Source.h

```diff
@@ -0,0 +1,161 @@
+/** @file Source.h
+  
+    @author Ananya Bahadur
+*/
+
+#ifndef CL_SOURCE_H
+#define CL_SOURCE_H
+
+#if ARDUINO >= 100
+ #include <Arduino.h>
+#else
+ #include <WProgram.h>
+#endif
+
+#include "Cyclops.h"
+
+/** @typedef operationMode */
+typedef enum {
+    LOOPBACK,
+    ONE_SHOT,
+} operationMode;
+
+/** @typedef sourceStatus */
+typedef enum {
+    FROZEN,
+    ACTIVE
+} sourceStatus;
+
+#define ONE_SHOT_FINISHED_HOLD_TIME 1000 /**< The waveform is completed.  
+                                          *   The timer would still interrupt
+                                          *    for this Waveform, as we don't
+                                          *    "pause" the waveform*.
+                                          *    Use a long HOLD_TIME.
+                                          */
+
+/**
+ * @brief 
+ * Source is an abstract class, representing a source of Voltage Data (as a 
+ * number between 0 and 4095) and the corresponding Time Data (as msec) for 
+ * which the Voltage should be held.
+ * 
+ * @details
+ * Multiple instances of a derived class can co-exist, they will be completely
+ * independent apart from the shared data source or technique.             
+ */
+class Source{
+ protected:
+    /** Keeps track of the number of SPI updates missed. When updates are
+     *  missed, the waveform remains stagnant. Since "time" doesn't stop,
+     *  Cyclops should behave as if it performed the SPI write. If this is not
+     *  done, this waveform would "lag" behind the intended waveform, and this 
+     *  lag would accumulate until FROZEN or reset.
+    */
+    uint8_t shift_accumulator;
+ public:
+    static uint8_t src_count;
+    /** @brief
+     *  Mode of operation. All derived sources support all variants.
+     *  @details
+     *  ``ONE_SHOT``: creates the waveform once and then gets FROZEN.
+     *  
+     *  ``LOOPBACK``: loops back to the starting to create a periodic waveform.
+    */
+    operationMode opMode;
+    /** @brief
+     *  State of the Source
+     *  @details
+     *  In ``FROZEN`` state stepForward() calls do nothing.
+    */    
+    sourceStatus status;
+    /**< Unique uint ID for this object */
+    const uint8_t src_id;
+
+    Source(operationMode opMode, sourceStatus = ACTIVE);
+
+    /**
+     * @brief      Returns the next data-point's Voltage.
+     */
+    virtual uint16_t nextVoltage() = 0;
+    /**
+     * @brief      Returns the next data-point's Hold-Time. The voltage will be held on the LED for this much time.
+     */
+    virtual uint16_t holdTime() = 0;
+    /**
+     * @brief      Moves to the next data-point.
+     * 
+     * @param[in]   num_of_steps Moves `num_of_steps` ahead into the waveform.
+     */
+    virtual void stepForward(uint8_t num_of_steps) = 0;
```

Comment:
Any reason to make `num_of_steps` limited to a small number (by declaring as `uint8_t`) other than it 'feels' that we should need to traverse so many points of the waveform?